sdk/php/src/Command/EntrypointCommand.php

```diff
@@ -0,0 +1,239 @@
+
+        $src = (new FindsSrcDirectory())();
+        $daggerObjects = (new FindsDaggerObjects())($src);
+
+        foreach ($daggerObjects as $daggerObject) {
+            $objectTypeDef = $this->daggerClient
+                ->typeDef()
+                ->withObject($this->normalizeClassname($daggerObject->name));
+
+            foreach ($daggerObject->daggerFunctions as $daggerFunction) {
+                $func = $this->daggerClient->function(
+                    $daggerFunction->name,
+                    $this->getTypeDef($daggerFunction->returnType)
+                );
+
+                if ($daggerFunction->description !== null) {
+                    $func = $func->withDescription($daggerFunction->description);
+                }
+
+                foreach ($daggerFunction->arguments as $argument) {
+                    $func = $func->withArg(
+                        $argument->name,
+                        $this->getTypeDef($argument->type),
+                        $argument->description,
+                        $argument->default
+                    );
+                }
+
+                $objectTypeDef = $objectTypeDef->withFunction($func);
+            }
+
+            $daggerModule = $daggerModule->withObject($objectTypeDef);
+        }
+
+        $functionCall->returnValue(new Dagger\Json(json_encode(
+            (string) $daggerModule->id()
+        )));
+
+        return Command::SUCCESS;
+    }
+
+    private function callFunctionOnParent(
+        OutputInterface $output,
+        Dagger\FunctionCall $functionCall,
+        string $parentName
+    ): int {
+        $errorOutput = $output instanceof ConsoleOutputInterface ?
+            $output->getErrorOutput() :
+            $output;
+
+        $className = "DaggerModule\\$parentName";
+        $functionName = $functionCall->name();
+        $class = new $className();
+        $class->client = $this->daggerClient;
+
+        $args = $this->formatArguments(
+            $className,
+            $functionName,
+            json_decode(json_encode($functionCall->inputArgs()), true)
+        );
+
+        try {
+            $result = ($class)->$functionName(...$args);
+        } catch (QueryError $e) {
+            if (!isset($e->getErrorDetails()['extensions'])) {
+                throw $e;
+            }
+
+            $errorOutput->writeln($e->getMessage());
+            $output->writeln($e->getErrorDetails()['extensions']['stdout'] ?? '');
+            $errorOutput->writeln($e->getErrorDetails()['extensions']['stderr'] ?? '');
+
+            return $e->getErrorDetails()['extensions']['exitCode'] ??
+                Command::FAILURE;
+        }
+
+        if ($result instanceof Dagger\Client\IdAble) {
+            $result = (string) $result->id();
+        }
+
+        $functionCall->returnValue(new Dagger\Json(json_encode($result)));
+
+        return Command::SUCCESS;
+    }
+
+    private function getTypeDef(Type $type): TypeDef
```

Comment:
If a type accepts null values, or has a default but you can't serialize it or report it to Dagger, you should set `$typeDef->withOptional(true)`.

Enums and custom scalars are easy to support too:

https://github.com/dagger/dagger/blob/aef8596314ad777085f11580c6911346f57c806c/sdk/python/src/dagger/mod/_converter.py#L111-L115

Examples you can try from core:
- Scalar: `Platform`
- Enum: `NetworkProtocol`

Reply:
Both should now be supported within this PR, thank you for the assistance!